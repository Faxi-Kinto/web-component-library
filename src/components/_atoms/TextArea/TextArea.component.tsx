import React, { useState, ChangeEvent, useMemo } from 'react';
import * as Styled from './TextArea.styles';
import Label from '../Label';
import classNames from 'classnames';

/**
 * @name TextArea
 *
 * @returns {JSX}
 */

export type TextAreaProps = {
  autoFocus?: boolean;
  className?: string;
  error?: boolean;
  horizontalResize?: boolean;
  label?: string;
  maxLength?: number;
  name?: string;
  noresize?: boolean;
  placeholder?: string;
  rows?: number;
  value?: string;
  verticalResize?: boolean;
  onChange?: (value: string) => void;
};

const TextArea: React.FC<TextAreaProps> = (
  props: TextAreaProps
): JSX.Element => {
  const {
    autoFocus,
    className,
    error = false,
    horizontalResize = false,
    label,
    maxLength,
    name,
    noresize = false,
    placeholder,
    rows,
    value,
    verticalResize = false,
    onChange,
    ...rest
  } = props;
  const [inputValue, setTextAreaValue] = useState<string>('');

  const htmlFor = useMemo(() => {
    if (name)
      return {
        htmlFor: name,
      };
    return {};
  }, [name]);

  const id = useMemo(() => {
    if (name)
      return {
        id: name,
        name,
      };
    return {};
  }, [name]);

  const handleOnChange = (event: ChangeEvent<HTMLTextAreaElement>): void => {
    if (onChange) {
      onChange(event.target.value);
    } else {
      setTextAreaValue(event.target.value);
    }
  };

  return (
    <Styled.TextAreaStyled>
      {label && <Label {...htmlFor}>{label}</Label>}
      <textarea
        autoFocus={autoFocus}
        className={classNames([
          'textarea',
          { 'textarea--error': error },
          { 'textarea--noresize': noresize },
          { 'textarea--horizontal-resize': horizontalResize },
          { 'textarea--vertical-resize': verticalResize },
          className,
        ])}
        maxLength={maxLength}
        placeholder={placeholder}
        rows={rows}
        value={value || inputValue}
        {...id}
        {...rest}
        onChange={handleOnChange}
      />
    </Styled.TextAreaStyled>
  );
};

export default TextArea;
