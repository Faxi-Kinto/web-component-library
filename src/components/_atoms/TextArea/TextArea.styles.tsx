import styled from 'styled-components';
import { marginChildren, pxToRem, padding } from '@faxi/web-css-utilities';

export const TextAreaStyled = styled.div`
  width: 100%;
  ${marginChildren(`0 0 ${pxToRem('10px')} 0`)};

  > textarea {
    width: 100%;
    height: ${pxToRem('45px')};
    ${padding(pxToRem('8px'), pxToRem('16px'))};
    outline: none;

    &:focus,
    &.textarea--error {
      box-shadow: none;
    }
  }
`;
