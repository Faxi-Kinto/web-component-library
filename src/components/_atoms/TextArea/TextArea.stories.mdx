import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import TextArea from './';

<Meta title="Core|Atoms/TextArea" />

# TextArea Component

## Implementation

#### Default

<Preview>
  <Story name="default" decorators={[withKnobs]}>
    <TextArea
      error={boolean('Error', false)}
      placeholder={text('Placeholder text', '')}
    />
  </Story>
</Preview>

#### With Error

<Preview>
  <Story name="error">
    <TextArea error={true} />
  </Story>
</Preview>

#### With Placeholder

<Preview>
  <Story name="placeholder">
    <TextArea placeholder="Type value here" error={false} />
  </Story>
</Preview>

#### With Label

<Preview>
  <Story name="label">
    <TextArea label="With label" name="labeled input" error={false} />
  </Story>
</Preview>

#### With Both Placeholder & Label

<Preview>
  <Story name="placeholder and label">
    <TextArea
      label="With label"
      placeholder="and also placeholder"
      name="test input"
      error={false}
    />
  </Story>
</Preview>

## TextArea Props

<Props of={TextArea} />

## Notes

- Default error value is set to false, so if omitted the field will render without error border

- With onChange prop you can set the state to the inputField from parent components.

- onnHandleCapslook prop is used for the password molecule, to inform component when caps-lock is pressed.
