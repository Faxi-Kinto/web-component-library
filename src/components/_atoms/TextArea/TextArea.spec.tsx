import React from 'react';
import renderer from 'react-test-renderer';
import TextArea from './';

test('TextArea renders correctly', () => {
  const component = renderer.create(
    <TextArea value="Test value" onChange={(): null => null} error />
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
