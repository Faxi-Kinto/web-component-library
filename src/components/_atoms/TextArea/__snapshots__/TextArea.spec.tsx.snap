// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
.c0 {
  width: 100%;
}

.c0 > * {
  margin: 0 0 0.625rem 0;
}

.c0 > :last-child {
  margin: 0;
}

.c0 > textarea {
  width: 100%;
  height: 2.8125rem;
  padding: 0.5rem 1rem 0.5rem 1rem;
  outline: none;
  font-family: inherit;
}

.c0 > textarea.textarea--noresize {
  resize: none;
}

.c0 > textarea.textarea--error {
  border: 0.03125rem solid #eb4820;
}

<div
  className="c0"
>
  <textarea
    className="textarea textarea--error"
    onChange={[Function]}
    value="Test value"
  />
</div>
`;
